### PR TITLE
Handle value-less parameters in `proj4_dict_to_str`

### PR DESCRIPTION
**Pull request by ethantkoenig**
Fix bug in `AreaDefinition.proj_str` where value-less parameters are given a value of `None`:

```
>>> ad = pyresample.AreaDefinition(
...     area_id='area_id',
...     description='description',
...     proj_id='tasmania_utm',
...     projection='+ellps=WGS84 +no_defs +proj=utm +south +towgs84=0,0,0 +type=crs +units=m +zone=50',
...     area_extent=(701609.0, 5127659.0, 901609.0, 5327659.0),
...     height=10,
...     width=10,
... )
>>> 
>>> ad.proj_str
(path to venv)/lib/python3.11/site-packages/pyproj/crs/crs.py:1286: UserWarning: You will likely lose important projection information when converting to a PROJ string from another format. See: https://proj.org/faq.html#what-is-the-best-format-for-describing-coordinate-reference-systems
  proj = self._crs.to_proj4(version=version)
'+ellps=WGS84 +no_defs +proj=utm +south=None +type=crs +units=m +zone=50'
```

 - [ x ] Tests added 
 - [ x ] Tests passed 
 - [ x ] Passes ``git diff origin/main **/*py | flake8 --diff``

### CODE REVIEW

**Comment by djhoese** [comment]
Thanks for working on this. In old versions of pyproj we did all the dict to string handling ourselves, but now we use pyproj's CRS object a lot more.

Does this match what the `pyproj` library does with `CRS.to_proj4()`? Actually, how do you feel about just using pyproj's `CRS` here? In future versions of pyresample and with semi-recent updates already, we've been phasing out the direct use of PROJ.4 in pyresample internals. It would be best to depend on pyproj and the pyproj CRS object as much as possible.

Thoughts?

**Comment by ethantkoenig** [comment]
> Does this match what the `pyproj` library does with `CRS.to_proj4()`?

`CRS.to_proj4()` [delegates to `proj_as_proj_string`](https://github.com/pyproj4/pyproj/blob/f5f85d3a22b90e41adfaf6bdf294185da610e3be/pyproj/_crs.pyx#L170-L175). I haven't look at the PROJ implementation of `proj_as_proj_string` in detail, but I'm reasonably confident it does not append `=None` :)

> Actually, how do you feel about just using pyproj's `CRS` here?

`CRS` doesn't seem to support sorted strings, so it can't be used for `proj4_dict_to_str` calls where `sort` is true.

**Comment by djhoese** [comment]
> CRS.to_proj4() [delegates to proj_as_proj_string](https://github.com/pyproj4/pyproj/blob/f5f85d3a22b90e41adfaf6bdf294185da610e3be/pyproj/_crs.pyx#L170-L175). I haven't look at the PROJ implementation of proj_as_proj_string in detail, but I'm reasonably confident it does not append =None :)

Haha. I think you're right, but I think it does do None or `"null"` when it creates a dict of PROJ.4 parameters. The dict case is I think why pyresample does it.

> CRS doesn't seem to support sorted strings, so it can't be used for proj4_dict_to_str calls where sort is true.

Ah good catch. Do you use `sort=True` in your own work/use case? This used to be (if I remember correctly) to help with hashing/caching and equality checks, but we depend completely on pyproj/PROJ for that now through the `CRS` objects.

I'd be willing to deprecate the `sort` option and/or push this functionality upstream if at all possible into pyproj or PROJ. I have a feeling though that those maintainers would prefer not to mess with PROJ.4 strings anymore.

@mraspaud @pnuu what do you think about deprecating the ability to sort PROJ.4 strings alphabetically?

**Comment by mraspaud** [comment]
I'm guessing the alphabetical sorting what to help comparing proj dicts in python2, so I don't mind removing that possibility

File: pyresample/test/test_utils.py [hunk 1]
```diff
@@ -198,6 +198,12 @@ def test_proj4_str_dict_conversion(self):
         self.assertIsInstance(proj_dict['lat_1'], float)
         self.assertIsInstance(proj_dict2['lat_1'], float)
 
+        # Value-less south parameter
+        proj_str = "+ellps=WGS84 +no_defs +proj=utm +south +type=crs +units=m +zone=54"
+        proj_dict = utils.proj4.proj4_str_to_dict(proj_str)
+        proj_str2 = utils.proj4.proj4_dict_to_str(proj_dict, sort=True)
+        self.assertEqual(proj_str, proj_str2)
```
Comment (mraspaud) [inline pyresample/test/test_utils.py:205]: Could this be put in a separate test function/method?
Also, I don't remember the code that is exercised here, but could you summarize how this test makes sure None value is given to `south`?

Comment (ethantkoenig) [reply pyresample/test/test_utils.py]: > Could this be put in a separate test function/method?

Done

> Also, I don't remember the code that is exercised here, but could you summarize how this test makes sure None value is given to `south`?

The test didn't check for the value associated with `south` in the dict (but was merely checking a `proj4_str_to_dict` -> `proj4_dict_to-str` round trip), but I added explicit assertions so now it does.

**Comment by djhoese** [comment]
I think I'd prefer maybe if we right tests that compare the PROJ `dict` rather than 2 strings that depend on order and could be wrong if pyproj adds additional arguments. What do you two (@ethantkoenig @mraspaud) think? Otherwise, what did we think about removing the sorting functionality and just depending on pyproj to do the conversion?

**Comment by ethantkoenig** [comment]
> I think I'd prefer maybe if we right tests that compare the PROJ `dict` rather than 2 strings that depend on order and could be wrong if pyproj adds additional arguments. What do you two (@ethantkoenig @mraspaud) think?

Sure, that works for me. Done.

**Comment by djhoese** [comment]
I've added this to the milestone for the next minor (feature) release of pyresample (I'll do a bug fix release before this release is worked on). I think I'd like this PR to evolve to be as I described above where pyproj is used for the conversion to a string. I think `sort` functionality should be deprecated and a warning added if it is specified:

```python
warnings.warn("The 'sort' option is no longer supported. Please use the comparison of pyproj 'CRS' objects instead.")
```

I'd be OK with completely removing the functionality so `sort` no longer does anything (has no effect) until it is completely removed. I'm 95% sure pyresample internals is the only place where `sort` *used* to be used.

@mraspaud @ethantkoenig thoughts?

**Comment by mraspaud** [comment]
just to confirm, I don't remember this arg being used anywhere else.

**Comment by mraspaud** [comment]
To clarify @djhoese do you want this PR to use pyproj, or can we merge this and make the pyproj version in a following PR?

**Comment by djhoese** [comment]
> To clarify @djhoese do you want this PR to use pyproj, or can we merge this and make the pyproj version in a following PR?

It wasn't clear to me from past comments (or maybe I'm forgetting or missed it) if @ethantkoenig was unable to do something in their own work because of the way `value-less` parameters are handled. If there is any type of semi-urgent need for this then this PR as-is can be merged. Otherwise, I don't see a reason to not switch to pyproj. With no urgency I see two options:

1. Remove all sorted behavior and switch completely to pyproj for this conversion. This is the harshest option and will break anyone's code that was dependent on `sort`. But also I'd want to know who is using it as soon as possible.
2. Add deprecation warning if `sort=True` is passed but still use the existing version of the function (with the fixes already suggested by @ethantkoenig in this PR for value-less parameters) but otherwise (`sort=False`) use pyproj for the conversion.

I suppose there is an inbetween option where `sort=True` warns but also has no effect.

**Comment by ethantkoenig** [comment]
> It wasn't clear to me from past comments (or maybe I'm forgetting or missed it) if @ethantkoenig was unable to do something in their own work because of the way `value-less` parameters are handled. If there is any type of semi-urgent need for this then this PR as-is can be merged. Otherwise, I don't see a reason to not switch to pyproj. With no urgency I see two options:

I have a slight preference for submitting the bug fix as-is and pursuing a long-term solution in a separate PR. But I'm also willing to start the envisioned pyproj refactor in this PR if you feel strongly about marrying the two.

> 
> 1. Remove all sorted behavior and switch completely to pyproj for this conversion. This is the harshest option and will break anyone's code that was dependent on `sort`. But also I'd want to know who is using it as soon as possible.
> 2. Add deprecation warning if `sort=True` is passed but still use the existing version of the function (with the fixes already suggested by @ethantkoenig in this PR for value-less parameters) but otherwise (`sort=False`) use pyproj for the conversion.
> 
> I suppose there is an inbetween option where `sort=True` warns but also has no effect.

I don't have a strong preference among any of these options (unless auditing whether anyone actually uses `sort` for option 1 would take time, in which case it'd be nice to have a quicker short-term fix).

File: pyresample/utils/proj4.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def proj4_dict_to_str(proj4_dict, sort=False):
     params = []
     for key, val in items:
         key = str(key) if key.startswith('+') else '+' + str(key)
-        if key in ['+no_defs', '+no_off', '+no_rot']:
+        if key in ['+no_defs', '+no_off', '+no_rot'] or val is None:
```
Comment (djhoese) [inline pyresample/utils/proj4.py:67]: I just realized, is there any point to the `key in [...]` check anymore if we assume that a value will be `None` for value-less parameters?

**Review by djhoese — APPROVED** [review]
I'm fine merging this as a simple one line change. I did realize something about the change and have commented about it inline. Let me know what you think. We can merge after we think about this. Hopefully another PR can do the rest of the work shortly after.

File: pyresample/utils/proj4.py (hunk 2 above)
Comment (ethantkoenig) [reply pyresample/utils/proj4.py:67]: I guess it depends on whether we assume that `no_{defs,off,rot}` keys will always have a value of `None` in `proj4_dict`.

I would think this is a safe assumption, but leaned towards keeping the `key in ...` check out of caution.

Comment (djhoese) [reply pyresample/utils/proj4.py:67]: Ok, let's keep it as is. I think it would be save to drop it, but given the plans to drop this altogether I'm not sure there is a point. Anyway, pyproj puts a `None` in the dict for `no_defs` so it seems like a safe assumption.
